lib/mqtt.js

```diff
@@ -4,9 +4,83 @@ var es = require('event-stream');
 var Long = require('long');
 var decoders = require('sgvdata/lib/protobuf');
 var direction = require('sgvdata/lib/utils').direction;
-var mqtt = require('mqtt');
 var moment = require('moment');
 
+function init (env, ctx) {
+
+  function mqtt ( ) {
+    return mqtt;
+  }
+
+  mqtt.client = connect(env);
+  var downloads = downloader();
+
+  if (mqtt.client) {
+    listenForMessages(ctx);
+  }
+
+  mqtt.every = every;
+  mqtt.entries = process();
+
+  //expose for tests that don't need to connect
+  mqtt.sgvSensorMerge = sgvSensorMerge;
+
+  function listenForMessages ( ) {
+    mqtt.client.on('message', function (topic, msg) {
+      console.log('topic', topic);
+      console.log(topic, 'on message', 'msg', msg.length);
+      switch (topic) {
+        case '/uploader':
+          console.log({type: topic, msg: msg.toString()});
+          break;
+        case '/downloads/protobuf':
+          downloadProtobuf(msg, topic, downloads, ctx);
+          break;
+
+        default:
+          console.log(topic, 'on message', 'msg', msg);
+          // ctx.entries.write(msg);
+          break;
+      }
+    });
+  }
+
+  mqtt.emitNotification = function emitNotification(notify) {
+    console.info('Publishing notification to mqtt: ', notify);
+    mqtt.client.publish('/notifications/json', JSON.stringify(notify), function mqttCallback (err) {
+      if (err) {
+        console.error('Unable to publish notification to MQTT', err);
+      }
+    });
+  };
+
+  return mqtt();
+}
+
+function connect (env) {
+  var uri = env.MQTT_MONITOR;
+  if (!uri) {
+    return null;
+  }
+
+  var opts = {
+    encoding: 'binary',
+    clean: false,
+    clientId: env.mqtt_client_id
+  };
+  var client = require('mqtt').connect(uri, opts);
+
+  function granted () { console.log('granted', arguments); }
+
+  client.subscribe('sgvs');
+  client.subscribe('published');
+  client.subscribe('/downloads/protobuf', {qos: 2}, granted);
+  client.subscribe('/uploader', granted);
+  client.subscribe('/entries/sgv', granted);
+
+  return client;
+}
+
 function process ( ) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: ['process' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/768406448/issues/source?bid=2452379&fileBranchId=2457009#l84)